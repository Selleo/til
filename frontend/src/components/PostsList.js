import React, { useEffect } from 'react'
import { useParams } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'

import { saveCategoryPosts } from '../store/actions/actions'
import { delayStep } from './Transition'
import Post from './Post'

import { statusType } from '../utils/constants'
import '../App.css'

const PostsList = ({ withCategory = false }) => {
  const categories = useSelector(state => state.categories)
  const categoryPosts = useSelector(state => state.categoryPosts?.posts)
  const posts = useSelector(state => state.posts)
  const categoriesStatus = useSelector(state => state.statuses.categoryPosts)
  const postStatus = useSelector(state => state.statuses.posts)

  const { id } = useParams()
  const dispatch = useDispatch()

  useEffect(() => {
    if (withCategory) {
      const foundCategory = categories.find(({ name }) => name === id)
      foundCategory && dispatch(saveCategoryPosts(foundCategory.id))
    }
  }, [withCategory, categories, dispatch, id])

  if (!withCategory && postStatus !== statusType.fetched) return null
  if (!!withCategory && categoriesStatus !== statusType.fetched) return null

  let delay = 0

  return withCategory
    ? categoryPosts.data.map(post => {
        delay += delayStep

        return <Post key={post.id} post={post} animationDelay={delay} />
      })
    : posts.data.map(post => {
        delay += delayStep

        return <Post key={post.id} post={post} animationDelay={delay} />
      })
}

export default PostsList
