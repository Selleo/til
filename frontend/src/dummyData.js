export const initialState = {
  categories: [],
  currentUser: null,
  users: [],
  posts: []
};

export const user = {
  email: "l.walczak@selleo.com",
  firstName: null,
  image:
    "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
  lastName: null,
  uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
};

export const allUsers = [
  {
    email: "l.walczak@selleo.com",
    firstName: null,
    image:
      "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
    lastName: null,
    uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
  },
  {
    email: "maciekPlacek@selleo.com",
    firstName: null,
    image:
      "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAIdsdfjdkjdfi/photo.jpg",
    lastName: null,
    uuid: "36528600-e215-49bf-8d5b-dadlfj33343443"
  }
];

export const categories = [
  { id: 1, name: "java" },
  { id: 2, name: "javascript" }
];

export const posts = [
  {
    author: {
      email: "l.walczak@selleo.com",
      firstName: null,
      image:
        "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
      lastName: null,
      uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
    },
    body: "adsfdf",
    categoriesIds: [2, 1],
    id: 1,
    isPublic: false,
    likes: [
      {
        post_id: 1,
        user_uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
      }
    ],
    likesCount: 1,
    title: "title "
  },
  {
    author: {
      email: "l.walczak@selleo.com",
      firstName: null,
      image:
        "https://lh5.googleusercontent.com/-4na6W1JJVa0/AAAAAAAAAAI/AAAAAAAAAAA/SDwbLxAgbqc/photo.jpg",
      lastName: null,
      uuid: "36528600-e215-49bf-8d5b-e1096b63c876"
    },
    body: "adfefsddf",
    categoriesIds: [3],
    id: 2,
    isPublic: false,
    likes: [],
    likesCount: 0,
    title: "maciekdfd"
  }
];
