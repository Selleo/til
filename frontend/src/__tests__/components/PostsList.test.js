import React from "react";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import configureMockStore from "redux-mock-store";
import { render } from "@testing-library/react";
import PostsList from "../../components/PostsList";
import { initialState, posts } from "../../dummyData";

describe("<PostsList/>", () => {
  const history = createMemoryHistory();
  const mockStore = configureMockStore([]);
  let store = mockStore(initialState);

  it("renders properly", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router history={history}>
          <PostsList />
        </Router>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe("when the posts are provided", () => {
    store = mockStore({ ...initialState, posts });

    it("renders list of posts", () => {
      const { getAllByTestId } = render(
        <Provider store={store}>
          <Router history={history}>
            <PostsList />
          </Router>
        </Provider>
      );

      expect(getAllByTestId("post-component").length).toEqual(2);
    });
  });
});
