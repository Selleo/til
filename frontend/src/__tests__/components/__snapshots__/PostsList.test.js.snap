// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostsList/> renders properly 1`] = `
<DocumentFragment>
  <div
    class="posts"
  >
    <section
      class="post"
      data-testid="post-component"
    >
      <article>
        <a
          href="/posts/1"
        >
          <h1>
            title 
          </h1>
        </a>
        <p>
          adsfdf
        </p>
      </article>
      <button>
        Copy post Url
      </button>
      <p>
        like: 0
      </p>
      <p>
        funny: 0
      </p>
      <p>
        love: 0
      </p>
      <p>
        surprised: 0
      </p>
    </section>
    <section
      class="post"
      data-testid="post-component"
    >
      <article>
        <a
          href="/posts/2"
        >
          <h1>
            maciekdfd
          </h1>
        </a>
        <p>
          adfefsddf
        </p>
      </article>
      <button>
        Copy post Url
      </button>
      <p>
        like: 0
      </p>
      <p>
        funny: 0
      </p>
      <p>
        love: 0
      </p>
      <p>
        surprised: 0
      </p>
    </section>
  </div>
</DocumentFragment>
`;
